### PR TITLE
Add mesh refinement for electrostatic solver and scraping

### PR DESCRIPTION
**Pull request by Yin-YinjianZhao**
**PR description edited by @RemiLehe**

This PR makes two functionalities compatible with mesh refinement:

- **Scraping:** This incorporates the changes from https://github.com/ECP-WarpX/WarpX/pull/2776. However, #2776 could not be tested in a full simulation script, because the electrostatic solver was not compatible with MR.

- **Electrostatic solver**: Only some of the Poisson solver that we are using (e.g. the one in the 3D non-EB version) can do multi-level solves, while the others (e.g. the one in the RZ and/or EB version) can only do single-level solves. In this PR, instead of doing a multi-level solve, we call the Poisson solver on a level-by-level basis. We start from the lower level (`lev = 0`) and interpolate the corresponding `phi` to the fine patch of the next level (`lev = 1`) (to provide both boundary conditions and an initial guess for the interior values). We then call the Poisson solver on the next level (`lev = 1`). We perform the same operations iteratively for the higher levels (`lev >= 2`).

### CODE REVIEW

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 1]
```diff
@@ -327,8 +328,60 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+            WarpXCommUtil::ParallelCopy(phi_cp, *phi[lev], 0, 0, 1, ng_src, ng, crse_period);
+
+            // Local interpolation from phi_cp to phi[lev+1]
+            // (Call mf_nodebilin_interp (from AMReX_MFInterp_C.H) in MFIter loop)
+            // (similar to https://github.com/ECP-WarpX/WarpX/blob/eeab8143e983df862b1f4646ea1baf6e6c85e33f/Source/Parallelization/WarpXComm.cpp#L329)
+            // but calling `mf_nodebilin_interp` instead of warpx_interp
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+            MultiFab dphi(phi_cp.boxArray(), phi_cp.DistributionMap(), phi_cp.nComp(), ng);
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 2]
```diff
@@ -327,8 +328,60 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+            for (MFIter mfi(*phi[lev]); mfi.isValid(); ++mfi)
+            {
+                Array4<Real> const& phi_aux = phi[lev]->array(mfi);
+                Array4<Real const> const& phi_fp = phi[lev]->const_array(mfi);
+                //Array4<Real const> const& phi_c = dphi.const_array(mfi);
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 3]
```diff
@@ -327,8 +328,60 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+        // Interpolation from phi[lev] to phi[lev+1]
+        // (This provides both the boundary conditions and initial guess for phi[lev+1])
+        if (lev < finest_level) {
+
+            // Allocate phi_cp for lev+1
+            // Similar to https://github.com/AMReX-Codes/amrex/blob/development/Src/LinearSolvers/MLMG/AMReX_MLMG.cpp#L576
+            // https://github.com/AMReX-Codes/amrex/blob/development/Src/LinearSolvers/MLMG/AMReX_MLMG.cpp#L587
+
+            MultiFab& phi_cp = *phi[lev+1];
+            BoxArray ba = phi_cp.boxArray();
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
            BoxArray ba = phi[lev+1]->boxArray();
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 4]
```diff
@@ -327,8 +328,60 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+            const IntVect& refratio = refRatio(lev-1);
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
            const IntVect& refratio = refRatio(lev);
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 5]
```diff
@@ -327,8 +328,60 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+            ba.coarsen(refratio);
+            const int ncomp = linop.getNComp();
+            int ng_dst = 1;
+            MultiFab cfine(ba, phi_cp.DistributionMap(), ncomp, ng_dst);
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
            MultiFab cfine(ba, phi[lev+1]->DistributionMap(), ncomp, 1);
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 6]
```diff
@@ -327,8 +328,60 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+
+            // Copy from phi[lev] to phi_cp (in parallel)
+            // (Similar to https://github.com/ECP-WarpX/WarpX/blob/eeab8143e983df862b1f4646ea1baf6e6c85e33f/Source/Parallelization/WarpXComm.cpp#L306)
+            const amrex::IntVect& ng_src = guard_cells.ng_FieldGather;
+            const amrex::Periodicity& crse_period = Geom(lev).periodicity();
+            const IntVect& ng = phi[lev]->nGrowVect();
+            WarpXCommUtil::ParallelCopy(phi_cp, *phi[lev], 0, 0, 1, ng_src, ng, crse_period);
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
            const amrex::Periodicity& crse_period = Geom(lev).periodicity();
            WarpXCommUtil::ParallelCopy(phi_cp, *phi[lev], 0, 0, 1, 1, 1, crse_period);
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 7]
```diff
@@ -327,8 +328,60 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+            // (Call mf_nodebilin_interp (from AMReX_MFInterp_C.H) in MFIter loop)
+            // (similar to https://github.com/ECP-WarpX/WarpX/blob/eeab8143e983df862b1f4646ea1baf6e6c85e33f/Source/Parallelization/WarpXComm.cpp#L329)
+            // but calling `mf_nodebilin_interp` instead of warpx_interp
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+            MultiFab dphi(phi_cp.boxArray(), phi_cp.DistributionMap(), phi_cp.nComp(), ng);
+            for (MFIter mfi(*phi[lev]); mfi.isValid(); ++mfi)
+            {
+                Array4<Real> const& phi_aux = phi[lev]->array(mfi);
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 8]
```diff
@@ -327,8 +328,60 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+                Array4<Real const> const& phi_fp = phi[lev]->const_array(mfi);
+                //Array4<Real const> const& phi_c = dphi.const_array(mfi);
+
+                //amrex::ParallelFor(Box(phi_aux),
+                //[=] AMREX_GPU_DEVICE (int j, int k, int l) noexcept
+                //{
+                //    warpx_interp(j, k, l, phi_aux, phi_fp, phi_c, phi_stag, refratio);
+                //});
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
                amrex::ParallelFor(Box(phi_aux),
                [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept
                {
                    mf_nodebilin_interp(i, j, k, 0, phi_fp_arr, 0, phi_cp_arr, 0, refratio);
                });
```

Comment (RemiLehe) [reply Source/FieldSolver/ElectrostaticSolver.cpp]: See this for more info:
https://amrex-codes.github.io/amrex/doxygen/namespaceamrex.html#ab0fde5a29bb528390db2710d71ea4825
Note that you will need to include `AMReX_MFInterp_C.H` at the top of this file, in order for `mf_nodebilin_interp` to be available.

**Review by RemiLehe — CHANGES_REQUESTED** [review]
Thanks for starting to work on this. 
I suggested a few changes ; could you include them in this PR?

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 9]
```diff
@@ -327,8 +328,53 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+
+            // Allocate phi_cp for lev+1
+            // Similar to https://github.com/AMReX-Codes/amrex/blob/development/Src/LinearSolvers/MLMG/AMReX_MLMG.cpp#L576
+            // https://github.com/AMReX-Codes/amrex/blob/development/Src/LinearSolvers/MLMG/AMReX_MLMG.cpp#L587
+
+            BoxArray ba = phi[lev+1]->boxArray();
+            const IntVect& refratio = refRatio(lev);
+            ba.coarsen(refratio);
+            const int ncomp = linop.getNComp();
+            MultiFab cfine(ba, phi[lev+1]->DistributionMap(), ncomp, 1);
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
            MultiFab phi_cp(ba, phi[lev+1]->DistributionMap(), ncomp, 1);
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 10]
```diff
@@ -327,8 +328,53 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+            // Local interpolation from phi_cp to phi[lev+1]
+            // (Call mf_nodebilin_interp (from AMReX_MFInterp_C.H) in MFIter loop)
+            // (similar to https://github.com/ECP-WarpX/WarpX/blob/eeab8143e983df862b1f4646ea1baf6e6c85e33f/Source/Parallelization/WarpXComm.cpp#L329)
+            // but calling `mf_nodebilin_interp` instead of warpx_interp
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+            for (MFIter mfi(*phi[lev]); mfi.isValid(); ++mfi)
+            {
+                Array4<Real const> const& phi_fp = phi[lev]->const_array(mfi);
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
                Array4<Real const> const& phi_fp_arr = phi[lev]->array(mfi);
                Array4<Real const> const& phi_cp_arr = phi_cp->array(mfi);                
```

**Review by RemiLehe — CHANGES_REQUESTED** [review]

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 11]
```diff
@@ -327,8 +329,55 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+            WarpXCommUtil::ParallelCopy(phi_cp, *phi[lev], 0, 0, 1, ng, ng, crse_period);
+
+            // Local interpolation from phi_cp to phi[lev+1]
+            // (Call mf_nodebilin_interp (from AMReX_MFInterp_C.H) in MFIter loop)
+            // (similar to https://github.com/ECP-WarpX/WarpX/blob/eeab8143e983df862b1f4646ea1baf6e6c85e33f/Source/Parallelization/WarpXComm.cpp#L329)
+            // but calling `mf_nodebilin_interp` instead of warpx_interp
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+            for (MFIter mfi(*phi[lev]); mfi.isValid(); ++mfi)
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
            for (MFIter mfi(*phi[lev+1],TilingIfNotGPU()); mfi.isValid(); ++mfi)
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 12]
```diff
@@ -327,8 +329,55 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
+            {
+                Array4<Real> const& phi_fp_arr = phi[lev+1]->array(mfi);
+                Array4<Real> const& phi_cp_arr = phi_cp.array(mfi);
+
+                amrex::ParallelFor(Box(phi[lev]->array(mfi)),
```
Comment (RemiLehe) [inline Source/FieldSolver/ElectrostaticSolver.cpp]:
```suggestion
                Box const& b = mfi.tilebox(phi[lev+1]->ixType().toIntVect());
                amrex::ParallelFor( b,
```

**Review by RemiLehe — CHANGES_REQUESTED** [review]

**Comment by Yin-YinjianZhao** [comment]
Right now, the PR can compile but has runtime errors as follows.
@RemiLehe and I couldn't figure out why, so we need help maybe from @WeiqunZhang and @atmyers.

1) A test in RZ with embedded BC and mesh refinement, based on `Examples/Tests/ElectrostaticSphereEB/inputs_rz`.
I compiled using `make -j 4 USE_RZ=TRUE USE_EB=TRUE DEBUG=TRUE`.
The input is [inputs_rz.txt](https://github.com/ECP-WarpX/WarpX/files/7943865/inputs_rz.txt).
The output is [outputs_rz.txt](https://github.com/ECP-WarpX/WarpX/files/7943868/outputs_rz.txt).
The Backtrace is [Backtrace_rz.txt](https://github.com/ECP-WarpX/WarpX/files/7943872/Backtrace_rz.txt)
In short, we obtained an out-of-bound error:
```
amrex::Abort::0:: (-1,10,0,0) is out of bound (-1:37,11:52,0:0,0:0) !!!
SIGABRT
```

2) A test in RZ with embedded BC and mesh refinement testing particle scarping, based on `Examples/Tests/scraping/inputs_rz`.
I compiled using `make -j 4 USE_RZ=TRUE USE_EB=TRUE DEBUG=TRUE`.
The input is [inputs_rz_scraping.txt](https://github.com/ECP-WarpX/WarpX/files/7943900/inputs_rz_scraping.txt).
The output is [outputs_rz_scraping.txt](https://github.com/ECP-WarpX/WarpX/files/7943902/outputs_rz_scraping.txt).
The Backtrace is [Backtrace_rz_scraping.txt](https://github.com/ECP-WarpX/WarpX/files/7943907/Backtrace_rz_scraping.txt).
In short, again we got this error:
```
amrex::Abort::amrex::Abort::0:: (-1,59,0,0) is out of bound (0:36,60:64,0:0,0:0) !!!
0:: (-1,-1,0,0) is out of bound (0:36,0:36,0:0,0:0) !!!
SIGABRT
```

3) A test in 3D with mesh refinement, based on `Examples/Tests/ElectrostaticSphere/inputs_3d`.
I compiled using `make -j 4 DEBUG=TRUE`.
The input is [inputs_3d.txt](https://github.com/ECP-WarpX/WarpX/files/7943922/inputs_3d.txt).
The output is [outputs_3d.txt](https://github.com/ECP-WarpX/WarpX/files/7943925/outputs_3d.txt).
The Backtrace is [Backtrace_3d.txt](https://github.com/ECP-WarpX/WarpX/files/7943926/Backtrace_3d.txt).
The error reads:
```
amrex::Abort::0:: (15,15,24,0) is out of bound (16:48,16:48,16:48,0:0) !!!
SIGABRT
amrex::Abort::0:: (15,15,40,0) is out of bound (16:48,16:48,16:48,0:0) !!!
amrex::Abort::0:: (15,15,32,0) is out of bound (16:48,16:48,16:48,0:0) !!!
amrex::Abort::0:: (15,15,15,0) is out of bound (16:48,16:48,16:48,0:0) !!!
```

**Comment by RemiLehe** [comment]
I think that the error in 1) probably comes from this line: 
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Parallelization/WarpXComm_K.H#L115
We probably need some drawing of the interpolation operation in order to understand why this is happening.

**Comment by RemiLehe** [comment]
After discussing with @EZoni , it seems that the above issues maybe independent from this PR.

While we confirm this, coud you finalize this PR? In particular, I think that the new tests require the corresponding benchmark to be set. (see https://warpx.readthedocs.io/en/latest/developers/checksum.html#reset-a-benchmark-from-a-plotfile-you-know-is-correct)

**Comment by RemiLehe** [comment]
Also: it seems that the `ElectrostaticSphereRZ` (i.e. `RZ` but without the embedded boundaries) failed to compile...

**Comment by EZoni** [comment]
I'll jump in with a question, triggered by another issue I was looking at recently. Do the boundary conditions defined in `ElectrostaticSolver::BoundaryHandler::definePhiBCs` need to be updated for more than one mesh refinement level as well?

**Comment by RemiLehe** [comment]
Regarding the problem 3: my impression is that is that it also occurs on the `development` branch (The traceback points to [this line](https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/WarpXParticleContainer.cpp#L726)). This should be reported as a separate issue. But for now, it can be ignored for this PR.

**Comment by Yin-YinjianZhao** [comment]
After the above merge, the test 3 mentioned here (https://github.com/ECP-WarpX/WarpX/pull/2728#issuecomment-1022434816) can run without any issues.

Test 2 still got an error. The output and the backtrace files are attached.

[Backtrace_rz_scraping2.txt](https://github.com/ECP-WarpX/WarpX/files/8091719/Backtrace_rz_scraping2.txt)
[outputs_rz_scraping2.txt](https://github.com/ECP-WarpX/WarpX/files/8091720/outputs_rz_scraping2.txt)

Besides, a print statement is added at https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/ParticleBoundaryBuffer.cpp#L230:
`std::cout << "lev: " << lev << " " << "(*distance_to_eb[lev])[pti].box(): " << (*distance_to_eb[lev])[pti].box() << std::endl;`, and we got:
```
STEP 1 starts ...
lev: 0 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (66,66) (1,1))
lev: 0 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (66,66) (1,1))
lev: 0 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (66,66) (1,1))
lev: 0 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (66,66) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,-2) (74,74) (1,1))
lev: 1 (*distance_to_eb[lev])[pti].box(): ((-2,118) (74,130) (1,1))
amrex::Abort::0:: (25,63,0,0) is out of bound (-2:74,118:130,0:0,0:0) !!!
SIGABRT
```

File: Source/WarpX.cpp [hunk 13]
```diff
@@ -1689,6 +1689,10 @@ WarpX::AllocLevelMFs (int lev, const BoxArray& ba, const DistributionMapping& dm
     Efield_avg_fp[lev][1] = std::make_unique<MultiFab>(amrex::convert(ba,Ey_nodal_flag),dm,ncomps,ngEB,tag("Efield_avg_fp[y]"));
     Efield_avg_fp[lev][2] = std::make_unique<MultiFab>(amrex::convert(ba,Ez_nodal_flag),dm,ncomps,ngEB,tag("Efield_avg_fp[z]"));
 
+    constexpr int nc_ls = 1;
+    constexpr int ng_ls = 2;
+    m_distance_to_eb[lev] = std::make_unique<MultiFab>(amrex::convert(ba, IntVect::TheNodeVector()), dm, nc_ls, ng_ls, tag("m_distance_to_eb"));
+
 #ifdef AMREX_USE_EB
```
Comment (RemiLehe) [inline Source/WarpX.cpp]:
```suggestion
#ifdef AMREX_USE_EB
    constexpr int nc_ls = 1;
    constexpr int ng_ls = 2;
    m_distance_to_eb[lev] = std::make_unique<MultiFab>(amrex::convert(ba, IntVect::TheNodeVector()), dm, nc_ls, ng_ls, tag("m_distance_to_eb"));

```

**Review by RemiLehe — APPROVED** [review]

**Comment by Yin-YinjianZhao** [comment]
Right now, it is found that the modified `Examples/Tests/scraping` test cannot pass after adding mesh refinement.
We found that the particles cannot be removed as desired and behave weird.
rz
rv
Without mesh refinement, the original scraping test gives:
rz
rv

**Comment by RemiLehe** [comment]
It looks like, for some particles, the `x` and `y` position jumps after the first iteration (but in a way that preserve the radius `r`). Here is for instance a plot where I tracked the `x`, `y` and `r` position of a particle.

Screen Shot 2022-02-22 at 3 25 04 PM

This suggests that maybe the `theta` information is not properly copied: https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/WarpXParticleContainer.H#L53

(As a side note, here is the code that was used to produce the plot - see [this tutorial](https://github.com/openPMD/openPMD-viewer/blob/dev/tutorials/4_Particle_selection.ipynb) for more info:
```
from openpmd_viewer import OpenPMDTimeSeries
from openpmd_viewer import ParticleTracker
import matplotlib.pyplot as plt
ts = OpenPMDTimeSeries('./diags/diag1/')
pt = ParticleTracker( ts, iteration=37, select={'z':[400e-3,450e-3]},  preserve_particle_index=True )
x, y, z = ts.iterate( ts.get_particle, ['x', 'y', 'z'], select=pt )

# index of the particle to look at
i = 100
plt.plot( ts.iterations, x[:,i], '-o', label='x');
plt.plot( ts.iterations, y[:,i], '-o', label='y');
plt.plot( ts.iterations, (x[:,i]**2 + y[:,i]**2)**.5, '-o', label='r');
plt.legend(loc=0)
plt.xlabel('Iteration')
```

**Comment by RemiLehe** [comment]
Btw, @Yin-YinjianZhao are you able to tell where the mesh refinement patches are located?
(You could e.g. plot them with `yt.annotate_grids()`.)

**Comment by RemiLehe** [comment]
@Yin-YinjianZhao Here are a few other things that @atmyers suggested to try:
- Could you try to run this case with a version of WarpX compiled *without EB*. In this case, the particles would of course not be removed at `r=0.1`, but it would be interesting to see whether we see the same problems in the first few iterations (i.e. some particles going in the wrong direction)
- Could you also try to run with `warpx.n_field_gather_buffer = 0` and `warpx.n_current_deposition_buffer = 0` in the input script, and see whether that fixes the issue?

**Comment by RemiLehe** [comment]
It seems that the bug comes from the fact that we are not sorting the angle theta of the particles here:
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/Sorting/Partition.cpp#L151

File: Regression/WarpX-tests.ini [hunk 14]
```diff
@@ -2479,6 +2479,23 @@ doVis = 0
 compareParticles = 0
 analysisRoutine = Examples/Tests/ElectrostaticSphereEB/analysis_rz.py
 
+[ElectrostaticSphereEB_RZ_MR]
+buildDir = .
+inputFile = Examples/Tests/ElectrostaticSphereEB/inputs_rz_mr
+runtime_params =
```
Comment (RemiLehe) [inline Regression/WarpX-tests.ini]:
```suggestion
runtime_params =  warpx.abort_on_warning_threshold = medium
```

**Comment by RemiLehe** [comment]
@Yin-YinjianZhao I attempted to fix the issue with [this commit](https://github.com/ECP-WarpX/WarpX/pull/2728/commits/562f1fbe1b81ede02713e9566aa1307751465384) above.
Could you re-compile and produce again the above small movie (with mesh refinement)?

**Comment by Yin-YinjianZhao** [comment]
@RemiLehe After your fix, I got this:
zr
rv

**Pull request merged**